### PR TITLE
activitywatch: init at 0.12.1

### PR DESCRIPTION
**Pull request by chaoky**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by auroraanna** [comment]
> maybe this should be called `activitywatch-bin`? as there is some effort to create a compiled version in #97379

nixpkgs doesn't have `-bin` stuff like Arch Linux for example. nixpkgs tries to make stuff build from source where ever possible and then Hydra builds it and puts it on cache.nixos.org. That cache will then contain the built package with the binary executable.

**Comment by chaoky** [comment]
Thanks for taking the time to clarify, so should this just be kept as `activitywatch`?
~ and updated with a compiled version when (and if) it's ever packaged from source

my question came from seeing a lot of `foo-bin` oss packages on nixpkgs
~aur builds are also [compiled and cache](https://aur.chaotic.cx/)~

**Comment by auroraanna** [comment]
Hmm, I'm not so sure anymore it it's unnescessary. `wkhtmltopdf` and `wkhtmltopdf-bin` exist in nixpkgs, same for Thunderbird.

**Comment by amz-x** [comment]
I'm also interested in this package. 

I do agree with @chaoky to rename it to `activitywatch-bin`, since the compiled version in #97379 would be `activitywatch` thus keeping the package naming convention consistent.

**Comment by unode** [comment]
@chaoky this recipe doesn't seem to work for me.
None of the bundled binaries have been patched.
They all still expect `/lib64/ld-linux-x86-64.so.2` to be the interpreter.

```
% file /nix/store/9cvi0pxc2dibqjsmryd5baz279888gc4-ActivityWatch-0.12.1/opt/activitywatch/aw-qt 
/nix/store/9cvi0pxc2dibqjsmryd5baz279888gc4-ActivityWatch-0.12.1/opt/activitywatch/aw-qt: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=9230f3fa19123436d7e9b46b02c11f1e492895c1, for GNU/Linux 2.6.32, stripped

% ./.nix-profile/bin/aw-qt 
zsh: no such file or directory: ./.nix-profile/bin/aw-qt
```

Did you do something else to have this recipe running in your system?

**Comment by chaoky** [comment]
> @chaoky this recipe doesn't seem to work for me. None of the bundled binaries have been patched. They all still expect `/lib64/ld-linux-x86-64.so.2` to be the interpreter.
> 
> ```
> % file /nix/store/9cvi0pxc2dibqjsmryd5baz279888gc4-ActivityWatch-0.12.1/opt/activitywatch/aw-qt 
> /nix/store/9cvi0pxc2dibqjsmryd5baz279888gc4-ActivityWatch-0.12.1/opt/activitywatch/aw-qt: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=9230f3fa19123436d7e9b46b02c11f1e492895c1, for GNU/Linux 2.6.32, stripped
> 
> % ./.nix-profile/bin/aw-qt 
> zsh: no such file or directory: ./.nix-profile/bin/aw-qt
> ```
> 
> Did you do something else to have this recipe running in your system?

It's wrapped now, the autoPatchelfHook is only there to pick up libz.so.1, I couldn't figure out how to include this library otherwise

**Comment by chaoky** [comment]
I see these on the runtime logs
```bash
/nix/store/8jlrffqj4h476lmd42l3flrpcl6z2813-ActivityWatch-0.12.1/opt/activitywatch/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /nix/store/3ph4rh9id7an858r1p7dj90nhf78g7s0-libproxy-0.4.18/lib/libproxy.so.1)
```
I'm sure weather this is a non-nixos issue, or how to fix it, if it isn't it. this comes from glib-networking, which is not essential for this package.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -39922,4 +39922,6 @@ with pkgs;
   zf = callPackage ../tools/misc/zf { };
 
   isolate = callPackage ../tools/security/isolate { };
+
+  activityWatch = callPackage ../applications/misc/activitywatch { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:39926]: Pleas sort this

File: pkgs/applications/misc/activitywatch/default.nix [hunk 2]
```diff
@@ -0,0 +1,49 @@
+{ config, lib, stdenv, fetchzip, qt6, wrapGAppsNoGuiHook, glib-networking, autoPatchelfHook }:
+
+stdenv.mkDerivation rec {
+  pname = "ActivityWatch";
+  version = "0.12.1";
+  meta = with lib; {
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/activitywatch/default.nix:6]: meta goes to the end of the file

File: pkgs/applications/misc/activitywatch/default.nix [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+  buildInputs = [ qt6.qtbase glib-networking ];
+  nativeBuildInputs = [ autoPatchelfHook wrapGAppsNoGuiHook ];
+  dontWrapQtApps = true;
+
+  installPhase = ''
+    installed=$out/opt/activitywatch
+    mkdir -p $installed
+    mv * $installed
+    mkdir -p $out/bin
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/activitywatch/default.nix:31]:
```suggestion
    mkdir -p $out/bin $out/share/applications $out/opt/activitywatch
    mv * $installed
```

File: pkgs/applications/misc/activitywatch/default.nix [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    # Symlink executables to /bin
+    ln -s $installed/aw-qt $out/bin/aw-qt
+    for name in "aw-server" "aw-server-rust" "aw-watcher-afk" "aw-watcher-window"; do
+        ln -s $installed/$name/$name $out/bin/$name
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/activitywatch/default.nix:36]:
```suggestion
      ln -s $installed/$name/$name $out/bin/$name
```
in nix code we use 2 spaces even for nested shellcode

File: pkgs/applications/misc/activitywatch/default.nix [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+  installPhase = ''
+    installed=$out/opt/activitywatch
+    mkdir -p $installed
+    mv * $installed
+    mkdir -p $out/bin
+
+    # Symlink executables to /bin
+    ln -s $installed/aw-qt $out/bin/aw-qt
+    for name in "aw-server" "aw-server-rust" "aw-watcher-afk" "aw-watcher-window"; do
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/activitywatch/default.nix:35]:
```suggestion
    for name in aw-server aw-server-rust aw-watcher-afk aw-watcher-window; do
```

File: pkgs/applications/misc/activitywatch/default.nix [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        ln -s $installed/$name/$name $out/bin/$name
+    done
+
+    # Add .desktop file to share/applications
+    mkdir -p $out/share/applications
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/activitywatch/default.nix:40]:
```suggestion
```

File: pkgs/applications/misc/activitywatch/default.nix [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    cp $installed/aw-qt.desktop $out/share/applications
+
+    # Add logo to share/icons
+    mkdir -p $out/share/icons/hicolor/512x512/apps
+    cp $installed/media/logo/logo.png $out/share/icons/hicolor/512x512/apps/activitywatch.png
+    mkdir -p $out/share/icons/hicolor/128x128/apps
+    cp $installed/media/logo/logo-128.png $out/share/icons/hicolor/128x128/apps/activitywatch.png
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/activitywatch/default.nix:47]:
```suggestion
    mkdir -p $out/share/icons/hicolor/{128x128,512x512}/apps
    ln -s $installed/media/logo/logo.png $out/share/icons/hicolor/512x512/apps/activitywatch.png
    ln -s $installed/media/logo/logo-128.png $out/share/icons/hicolor/128x128/apps/activitywatch.png
```

File: pkgs/applications/misc/activitywatch/default.nix [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    platforms = platforms.linux;
+  };
+
+  src = fetchzip {
+    url = "https://github.com/ActivityWatch/activitywatch/releases/download/v${version}/activitywatch-v${version}-linux-x86_64.zip";
+    hash = "sha256-VdnAKcLQBWBVZqRyD1xoYuAfvTeCtcFJp3/slkQ6ynk=";
+  };
+
+  buildInputs = [ qt6.qtbase glib-networking ];
+  nativeBuildInputs = [ autoPatchelfHook wrapGAppsNoGuiHook ];
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/activitywatch/default.nix:24]: Why not compile from source?

**Review by SuperSandro2000 — COMMENTED** [review]
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) when naming your commits.

File: pkgs/applications/misc/activitywatch/default.nix (hunk 8 above)
Comment (chaoky) [reply pkgs/applications/misc/activitywatch/default.nix:24]: activitywatch is composed of a lot of small services, the derivation proposed on the related issue is 380 lines long

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (chaoky) [reply pkgs/top-level/all-packages.nix]: sorted!

File: pkgs/applications/misc/activitywatch/default.nix [hunk 9]
```diff
@@ -0,0 +1,51 @@
+{ config, lib, stdenv, fetchzip, qt6, wrapGAppsNoGuiHook, glib-networking
+, autoPatchelfHook }:
+
+stdenv.mkDerivation rec {
+  pname = "ActivityWatch";
+  version = "0.12.1";
+
+  src = fetchzip {
+    url =
+      "https://github.com/ActivityWatch/activitywatch/releases/download/v${version}/activitywatch-v${version}-linux-x86_64.zip";
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/activitywatch/default.nix:10]:
```suggestion
    url = "https://github.com/ActivityWatch/activitywatch/releases/download/v${version}/activitywatch-v${version}-linux-x86_64.zip";
```

File: pkgs/applications/misc/activitywatch/default.nix (hunk 8 above)
Comment (SuperSandro2000) [reply pkgs/applications/misc/activitywatch/default.nix:15]: which is preferable if someone already has done the work

File: pkgs/applications/misc/activitywatch/default.nix [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+    # Add logo to share/icons
+    mkdir -p $out/share/icons/hicolor/{128x128,512x512}/apps
+    ln -s $installed/media/logo/logo.png $out/share/icons/hicolor/512x512/apps/activitywatch.png
+    ln -s $installed/media/logo/logo-128.png $out/share/icons/hicolor/128x128/apps/activitywatch.png
+  '';
+
+  meta = with lib; {
+    description =
+      "ActivityWatch is an app that automatically tracks how you spend time on your devices.";
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/activitywatch/default.nix:40]:
```suggestion
    description = "ActivityWatch is an app that automatically tracks how you spend time on your devices.";
```

**Comment by SuperSandro2000** [comment]
@ofborg eval

File: pkgs/applications/misc/activitywatch/default.nix (hunk 8 above)
Comment (chaoky) [reply pkgs/applications/misc/activitywatch/default.nix:14]: I'll try, what do you recommend for node packages? buildNpmPackage?

**Comment by unode** [comment]
Superseded by #202935

**Pull request closed without merging**